lab2/part2/myrouter_part2.py

```diff
@@ -0,0 +1,443 @@
+    Forwards an IPv4 packet to a new destination through a specific port
+      pkt       Packet to forward (Ethernet header will be replaced)
+      out_port  Output port
+      dst_mac   Destination MAC
+    '''
+    def forward_ipv4(self, pkt: Packet, out_port: EthAddr, dst_mac: EthAddr):
+        
+        #Create new Ethernet Header
+        del pkt[0]
+        eth_head = Ethernet(src=out_port, dst=dst_mac, ethertype=EtherType.IPv4)
+        pkt.prepend_header(eth_head)
+
+        #Decrement TTL
+        pkt[IPv4].ttl -= 1
+
+        self.send_packet(pkt, out_port)
+
+    '''
+    handle_arp
+    Handles an ARP header in the packet (if one exists)
+    Sends an ARP reply to requests that can be serviced
+    Updates the ArpContext table for replies to earlier requests
+      pkt           Packet to handle
+      input_port    port the packet arrived from
+    '''
+    def handle_arp(self, pkt: Packet, input_port: EthAddr):
+
+        arp_head = pkt.get_header(Arp)
+        if not isinstance(arp_head, Arp): return
+
+        if arp_head.targetprotoaddr in self.local_proto_eth:
+            if arp_head.operation == ArpOperation.Request:
+                self.send_packet(self.local_proto_eth.handle_arp_request(arp_head), input_port)
+                return None
+            
+            elif arp_head.operation == ArpOperation.Reply:
+                return self.other_proto_eth.handle_arp_reply(arp_head)
+
+    '''
+    send_packet
+    Sends a packet through a port
+      pkt       Packet to send
+      port      Port (MAC) to send the packet through
+    '''
+    def send_packet(self, pkt, port):
+        if not isinstance(pkt, Packet):return
+
+        log_debug("R: Send Packet: {} on {}".format(str(pkt), str(port)))
+
+        out_intf = self.net.port_by_macaddr(port)
+        try:
+            self.net.send_packet(out_intf, pkt)
+        except ValueError as e:
+            log_debug("Failed to send packet. Got ValueError: {}".format(e))
+        except:
+            log_debug("Failed to send packet. Unknown Error: {}".format(sys.exc_info()))
+
+    '''
+    enqueue_packet
+    Side-tracks a packet until its destination IP can be resolved to a MAC
+    Generates and sends an ARP requests, then places the packet on a queue
+      pkt           Packet
+      port          Output port
+      addr          Destination IPv4 addr
+    '''
+    def enqueue_packet(self, pkt: Packet, port: EthAddr, addr: IPv4Address):
+        #Generate ARP request
+        arp_pkt = self.local_proto_eth.get_arp_request(addr, self.net.interface_by_macaddr(port))
+        self.send_packet(arp_pkt, port)
+
+        #Enqueue the packet to be sent later
+        self.queue.append(Router.QueuedPacket(pkt, port, addr))
+
+    '''
+    dequeue_packets
+    Evaluates side-tracked packets. Sends packets with resolvable destinations.
+    For packets still waiting for destinations, sends up to 2 additional ARP requests
+    Drops packets after 3 seconds and 3 ARP requests.
```

Comment:
I think this comment should just be `Drops packets after 3 ARP requests`, because the **minimum** time to wait is 1 second before re-try; therefore, the `after 3 seconds` part cant be safely dropped from the comment.

And of course there is the caveat with the 3rd request that we still have to wait at last a full second before we can drop the item from the queue.

Reply:
Agreed - definitely clearer